File: 2024/Day-03/Part-Two.js
import {readFile} from 'node:fs/promises';

let fileContent = await readFile('input.txt', 'utf8');
let sumOfMultiplication = 0; // Final result
const regex = /mul\((\d{1,3}),(\d{1,3})\)/g; // Representation of 'mul(n, m)' in regex format : n & m are variable numbers of 3 digits maximum

while (fileContent.includes("don't()") ) {

    let index = getIndex(fileContent)
    fileContent = removeSubstring(fileContent, index[0], index[1])

}


const matches = [...fileContent.matchAll(regex)];
    matches.forEach((match)=>{
       let n = match[1]
       let m = match[2]
       sumOfMultiplication += parseInt(m, 10) * parseInt(n, 10)
    })  

console.log(sumOfMultiplication)

function getIndex(str) {
    let indexDont = undefined
    let indexDo = undefined
    for (let i = 0; i < str.length; i++) {
        if (str[i] === 'd' && str[i + 1] === 'o' && str[i + 2] === 'n' && str[i + 3] === "'" && str[i + 4] === 't' && str[i + 5] === '(' && str[i + 6] === ')') {
            indexDont = i
            break
        } 
    }
    for (let i = 0; i < str.length; i++) {    
        if( (indexDont < i ) && (str[i] === 'd' && str[i + 1] === 'o' && str[i + 2] ==='(' && str[i + 3] === ')') ) {
                indexDo = i
                break
            }
    }
    if (indexDo && indexDont)
        return [indexDont, indexDo]
    else if (!indexDo )
        return [indexDont]
}


function removeSubstring(str, begin, end = str.length) {
    if (begin < 0 || begin >= str.length || begin >= end) {
      throw new Error("Indices invalides !");
    }
    // Concaténer la partie avant 'begin' et celle après 'end'
    return str.slice(0, begin) + str.slice(end);
  }
  

File: 2015/Day-03/Part-Two.py
# Santa
x1, y1 = 0, 0
# Robo-Santa
x2, y2 = 0, 0  

house_list = {(0, 0)}
with open('input.txt') as file:
    for line in file:
        for index, el in enumerate(line):
            match el:
                case '>': 
                    dx, dy = 1, 0
                case '<': 
                    dx, dy = -1, 0
                case '^':
                    dx, dy = 0, 1
                case 'v': 
                    dx, dy = 0, -1
                case _:
                    continue

            if index % 2 == 0:
                x1 += dx
                y1 += dy
                house_list.add((x1, y1))
            else:
                x2 += dx
                y2 += dy
                house_list.add((x2, y2))

print(len(house_list))

